session/v2/token.go

```diff
@@ -0,0 +1,1011 @@
+	if x.version != TokenCurrentVersion {
+		return fmt.Errorf("invalid token version: expected %d, got %d", TokenCurrentVersion, x.version)
+	}
+
+	if x.issuer.IsZero() {
+		return errors.New("issuer is not set")
+	}
+
+	if len(x.subjects) == 0 {
+		return errors.New("no subjects specified")
+	}
+
+	if len(x.subjects) > MaxSubjectsPerToken {
+		return fmt.Errorf("too many subjects: expected max %d, got %d", MaxSubjectsPerToken, len(x.subjects))
+	}
+
+	for i, subj := range x.subjects {
+		if subj.IsZero() {
+			return fmt.Errorf("subject at index %d is empty", i)
+		}
+	}
+
+	if x.Iat().IsZero() {
+		return errors.New("issued at (iat) is not set")
+	}
+	if x.Nbf().IsZero() {
+		return errors.New("not valid before (nbf) is not set")
+	}
+	if x.Exp().IsZero() {
+		return errors.New("expiration (exp) is not set")
+	}
+	if x.Nbf().After(x.Exp()) {
+		return errors.New("not before (nbf) is after expiration (exp)")
+	}
+	if x.Iat().After(x.Exp()) {
+		return errors.New("issued at (iat) is after expiration (exp)")
+	}
+
+	if len(x.contexts) == 0 {
+		return errors.New("no contexts specified")
+	}
+
+	if len(x.contexts) > MaxContextsPerToken {
+		return fmt.Errorf("too many contexts: expected max %d, got %d", MaxContextsPerToken, len(x.contexts))
+	}
+
+	for i, ctx := range x.contexts {
+		if len(ctx.verbs) == 0 {
+			return fmt.Errorf("context at index %d has no verbs", i)
+		}
+		if len(ctx.verbs) > MaxVerbsPerContext {
+			return fmt.Errorf("context at index %d has too many verbs: expected max %d, got %d", i, MaxVerbsPerContext, len(ctx.verbs))
+		}
+		if len(ctx.objects) > MaxObjectsPerContext {
+			return fmt.Errorf("context at index %d has too many objects: expected max %d, got %d", i, MaxObjectsPerContext, len(ctx.objects))
+		}
+	}
+
+	if !x.sigSet {
+		return errors.New("token is not signed")
+	}
+	if !x.VerifySignature() {
+		return errors.New("token signature verification failed")
+	}
+
+	return nil
+}
+
+// validateDelegatedContexts checks that all verbs and containers in delegated token contexts
+// are authorized by origin token.
+func (x Token) validateDelegatedContexts() error {
+	originByContainer := x.indexContextsByContainer(x.origin.contexts)
+	hasWildcard := len(originByContainer[cid.ID{}]) > 0
+
+	delegatedByContainer := x.indexContextsByContainer(x.contexts)
+
+	for containerKey, delegatedContexts := range delegatedByContainer {
+		originContexts := originByContainer[containerKey]
+
+		// Container not found in origin
+		if len(originContexts) == 0 {
+			if hasWildcard {
+				originContexts = originByContainer[cid.ID{}]
+			} else {
+				continue
+			}
+		}
+
+		if err := x.validateContainerContexts(containerKey, delegatedContexts, originContexts); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// indexContextsByContainer creates a temporary map of contexts grouped by container ID for fast lookup.
+func (x Token) indexContextsByContainer(contexts []Context) map[cid.ID][]Context {
+	index := make(map[cid.ID][]Context)
```

Comment:
Avoid allocations during checks. It's pretty easy if you're to require inner tokens to have the same order of containers as outer ones. This is a minor update to the spec, but it allows to do checks sequentially without any additional constructs.